program/src/processor.rs

```diff
@@ -644,3 +644,117 @@ pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> P
+            // Evaluate.
+            if token_supply == 0 {
+                prop_assert_eq!(result, Ok(0));
+            } else {
+                // For all possible values of rewards and token_supply, the
+                // calculation should never return an error, hence the
+                // `unwrap` here.
+                //
+                // The scaling to `u128` prevents multiplication from breaking
+                // the `u64::MAX` ceiling, and the `token_supply == 0` check
+                // prevents `checked_div` returning `None` from a zero
+                // denominator.
+                let expected = (rewards as u128)
+                    .checked_mul(REWARDS_PER_TOKEN_SCALING_FACTOR)
+                    .and_then(|product| product.checked_div(token_supply as u128))
+                    .unwrap();
+                prop_assert_eq!(result, Ok(expected));
+            }
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn test_calculate_eligible_rewards(
+            current_accumulated_rewards_per_token in 0u128..,
+            last_accumulated_rewards_per_token in 0u128..,
+            token_account_balance in 0u64..,
+        ) {
+            // Calculate.
+            let result = calculate_eligible_rewards(
+                current_accumulated_rewards_per_token,
+                last_accumulated_rewards_per_token,
+                token_account_balance,
+            );
+            // Evaluate.
+            let marginal_rate = current_accumulated_rewards_per_token
+                .checked_sub(last_accumulated_rewards_per_token);
+            match marginal_rate {
+                Some(marginal_rate) => {
+                    if marginal_rate == 0 {
+                        // If the marginal rate resolves to zero, the
+                        // calculation should short-circuit and return zero.
+                        prop_assert_eq!(result, Ok(0));
+                    } else {
+                        // The rest of the calculation consists of three steps,
+                        // so evaluate each step one at a time.
+                        //
+                        // 1. marginal rate x token account balance
+                        // 2. product / REWARDS_PER_TOKEN_SCALING_FACTOR
+                        // 3. product.try_into (u64)
+
+                        // Step 1.
+                        let marginal_rewards = marginal_rate
+                            .checked_mul(token_account_balance as u128);
+                        if marginal_rewards.is_none() {
+                            // If the product of the marginal rate and the token
+                            // account balance overflows, the calculation should
+                            // return an arithmetic error.
+                            prop_assert_eq!(result, Err(ProgramError::ArithmeticOverflow));
+                            return Ok(());
+                        }
```

Comment:
Okay I ran some prop tests and played around with the inputs and shrinkage to define a few bounds, but I've also included the constraints of SOL supply and mint supply that guarantee this calculation's accuracy. I've included the extraneous bounds for documentation's sake, in case this program is forked and a mint with a different supply is used.